[clang][dep-scan] Resolve lexer crash from a permutation of invalid tokens

cyndyishida opened the pull request:
Sometimes, when a user writes invalid code, the minimization used for scanning can create a stream of tokens that is invalid at lex time. This patch protects against the case where there are valid (non-c++20) import directives discovered in the middle of an invalid `import` declaration.

Mostly authored by: @akyrtzi 
resolves: rdar://152335844

akyrtzi commented on clang/lib/Lex/DependencyDirectivesScanner.cpp:
```diff
@@ -503,6 +503,10 @@ bool Scanner::lexModuleDirectiveBody(DirectiveKind Kind, const char *&First,
     if (Tok.is(tok::semi))
       break;
+    if (Tok.is(tok::hash) || Tok.is(tok::at))
+      return reportError(
```
IMHO there's no benefit in emitting an error during minimization, unless there's a reason that the frontend cannot emit a proper diagnostic during compilation.
The scanner's primary focus should be on identifying dependencies; invalid code can be reported by the frontend.

akyrtzi replied:
The `#import` on its own line should be accepted (show up in minimization) I think?

cyndyishida replied:
I'm not really sure how to do that without introducing some backtracking. AFAICT, as we consume each token, the `First` keeps moving forward.

akyrtzi replied:
I think this should work fine for the test case:
```
@@ -496,7 +496,12 @@ bool Scanner::lexModuleDirectiveBody(DirectiveKind Kind, const char *&First,
                                      const char *const End) {
   const char *DirectiveLoc = Input.data() + CurDirToks.front().Offset;
   for (;;) {
+    const char *PriorFirst = First;
     const dependency_directives_scan::Token &Tok = lexToken(First, End);
+    if (Tok.is(tok::hash) && Tok.Flags & clang::Token::StartOfLine) {
+      First = PriorFirst;
+      return false;
+    }
     if (Tok.is(tok::eof))
       return reportError(
           DirectiveLoc,
```

cyndyishida replied:
Thank you for taking a look! Applied your suggested patch, with minor tweaks.

Bigcheese commented on clang/test/Lexer/minimize_source_to_dependency_directives_invalid_import.m:
```diff
@@ -0,0 +1,6 @@
+// RUN: %clang_cc1 -verify -print-dependency-directives-minimized-source %s 2>&1
+
+import <invalid.h>
+#import "invalid.h" // expected-error {{unexpected extra tokens inside declaration}}
```
This is valid for MSVC for COM related stuff and is also dependency relevant there. I think it would be best to just forward this to the frontend to deal with.

cyndyishida replied:
In what language mode? Ultimately, the minimized output does generate an assertion failure or a crash (depending on how the compiler was built) because the minimizer is attempting to reason about invalid code.
It's not very clear to me how to avoid this scenario in a way that appeases all expected use cases.

cyndyishida committed:
[clang][dep-scan] Resolve lexer crash from a permutation of invalid tokens

Sometimes, when a user writes invalid code, the minimization used for scanning can create a stream of tokens that is invalid at lex time. This patch protects against the case where there are valid (non-c++20) import directives discovered in the middle of an invalid import declaration.

resolves: rdar://152335844
